File: cli/src/ship/compose.ts
import { RESOURCES } from '../add/resources.js'
import type { ResourceRecord } from '../add/plan.js'
import type { ProdForm } from '../add/resources.js'

/**
 * Продакшен-половина ресурсов.
 *
 * Локально ресурсы — это команды docker run в конфиге; на сервере — службы в
 * одном compose-файле. Описание при этом одно: и то и другое собирается из
 * секции resources, которую написала команда add. Поэтому база в проде — та же
 * база, что на ноутбуке, а не её пересказ по памяти.
 */

/**
 * Минимальный вывод YAML.
 *
 * Полноценная библиотека здесь не нужна: мы печатаем свою же структуру, а не
 * разбираем чужую. Строки всегда в двойных кавычках — так не нужно гадать,
 * какие из них YAML примет за число, дату или значение вроде no.
 */
export function toYaml(value: unknown, indent = 0): string {
  const pad = '  '.repeat(indent)

  if (Array.isArray(value)) {
    if (value.length === 0) return ' []\n'
    return '\n' + value.map((v) => `${pad}- ${toYaml(v, indent + 1).trimStart()}`).join('')
  }

  if (value !== null && typeof value === 'object') {
    const entries = Object.entries(value as Record<string, unknown>)
    if (entries.length === 0) return ' {}\n'
    return '\n' + entries.map(([k, v]) => `${pad}${k}:${toYaml(v, indent + 1)}`).join('')
  }

  if (typeof value === 'string') return ` "${value.replace(/\\/g, '\\\\').replace(/"/g, '\\"')}"\n`
  return ` ${String(value)}\n`
}

export interface ComposePlan {
  yaml: string
  /** Файлы, которые едут в shared/ рядом с compose. */
  files: Record<string, string>
  /** Что должно оказаться в shared/.env на сервере. */
  env: Record<string, string>
  /** Ресурсы, которым в проде не место: mailpit и подобные. */
  skipped: string[]
  /** Имена служб в compose. */
  services: string[]
}

/**
 * Собрать compose из реестра ресурсов проекта.
 *
 * Ресурсы, которых нет в каталоге, пропускаются молча: конфиг мог быть написан
 * более новой версией инструмента, и падать из-за этого нельзя.
 */
export function planCompose(resources: Record<string, ResourceRecord>, project: string): ComposePlan | null {
  const services: Record<string, unknown> = {}
  const volumes: Record<string, unknown> = {}
  const files: Record<string, string> = {}
  const env: Record<string, string> = {}
  const skipped: string[] = []

  for (const [name, record] of Object.entries(resources)) {
    const resource = RESOURCES.find((r) => r.id === record.type)
    if (!resource) continue
    if (!resource.prod) {
      skipped.push(`${name} (${resource.title})`)
      continue
    }

    const form: ProdForm = resource.prod({
      name,
      container: `${project}-${name}`,
      project,
      ports: record.ports,
    })

    services[name] = form.service
    for (const v of form.volumes) volumes[v] = null
    Object.assign(env, form.env)
    Object.assign(files, form.files ?? {})
  }

  if (Object.keys(services).length === 0) return null

  const doc: Record<string, unknown> = { services }
  if (Object.keys(volumes).length > 0) doc.volumes = volumes

  const header =
    '# Собрано командой p3k ship из секции resources в p3k.json.\n' +
    '# Правки здесь потеряются при следующей выкатке — правьте конфиг проекта.\n' +
    '# Пароли берутся из shared/.env рядом с этим файлом и сюда не попадают.\n'

  return { yaml: header + toYaml(doc).trimStart(), files, env, skipped, services: Object.keys(services) }
}

/** Заготовка shared/.env: значения в угловых скобках человек заполняет сам. */
export function envExample(env: Record<string, string>): string {
  const lines = [
    '# Переменные окружения продакшена. Файл остаётся на сервере и никуда',
    '# не отправляется. Значения в угловых скобках замените своими.',
    '',
  ]
  for (const [key, value] of Object.entries(env)) lines.push(`${key}=${value}`)
  return lines.join('\n') + '\n'
}

/** Переменные, которые человек ещё не заполнил. */
export const unfilled = (env: Record<string, string>) =>
  Object.entries(env)
    .filter(([, v]) => v.includes('<'))
    .map(([k]) => k)

File: go/compose.go
package p3k

import (
	"fmt"
	"sort"
	"strings"
)

// ComposePlan — продакшен-половина ресурсов проекта.
//
// Локально ресурсы — это команды docker run в конфиге; на сервере — службы в
// одном compose-файле. Описание при этом одно: и то и другое собирается из
// секции resources. Поэтому база в проде — та же база, что на ноутбуке, а не
// её пересказ по памяти.
type ComposePlan struct {
	YAML string
	// Files — что едет в shared/ рядом с compose.
	Files map[string]string
	// Env — что должно оказаться в shared/.env на сервере.
	Env map[string]string
	// Skipped — ресурсы, которым в проде не место: mailpit и подобные.
	Skipped []string
	// Services — имена служб в compose.
	Services []string
}

// Compose собирает продакшен-манифест из реестра ресурсов проекта.
//
// Возвращает nil, если поднимать на сервере нечего: ресурсов нет вовсе или все
// они только для разработки. Ресурсы, которых нет в каталоге, пропускаются
// молча — конфиг мог быть написан более новой версией, и падать из-за этого
// нельзя.
func Compose(resources map[string]ResourceRef, project string) *ComposePlan {
	services := map[string]any{}
	volumes := map[string]any{}
	plan := &ComposePlan{Files: map[string]string{}, Env: map[string]string{}}

	names := make([]string, 0, len(resources))
	for name := range resources {
		names = append(names, name)
	}
	sort.Strings(names)

	for _, name := range names {
		ref := resources[name]
		def, ok := FindResource(ref.Type)
		if !ok {
			continue
		}
		if def.Prod == nil {
			plan.Skipped = append(plan.Skipped, fmt.Sprintf("%s (%s)", name, def.Title))
			continue
		}

		ports := def.DefaultPorts()
		for id, p := range ref.Ports {
			ports[id] = p
		}

		form := def.Prod(ResourceOpts{
			Name:      name,
			Container: project + "-" + name,
			Project:   project,
			Ports:     ports,
		})

		services[name] = form.Service
		plan.Services = append(plan.Services, name)
		for _, v := range form.Volumes {
			volumes[v] = nil
		}
		for k, v := range form.Env {
			plan.Env[k] = v
		}
		for k, v := range form.Files {
			plan.Files[k] = v
		}
	}

	if len(services) == 0 {
		return nil
	}

	doc := map[string]any{"services": services}
	if len(volumes) > 0 {
		doc["volumes"] = volumes
	}

	plan.YAML = "# Собрано библиотекой p3k из секции resources в p3k.json.\n" +
		"# Правки здесь потеряются при следующей выкатке — правьте конфиг проекта.\n" +
		"# Пароли берутся из shared/.env рядом с этим файлом и сюда не попадают.\n" +
		strings.TrimPrefix(toYAML(doc, 0), "\n")

	return plan
}

// EnvExample — заготовка shared/.env: значения в угловых скобках человек
// заполняет сам.
func (p *ComposePlan) EnvExample() string {
	lines := []string{
		"# Переменные окружения продакшена. Файл остаётся на сервере и никуда",
		"# не отправляется. Значения в угловых скобках замените своими.",
		"",
	}
	keys := make([]string, 0, len(p.Env))
	for k := range p.Env {
		keys = append(keys, k)
	}
	sort.Strings(keys)
	for _, k := range keys {
		lines = append(lines, k+"="+p.Env[k])
	}
	return strings.Join(lines, "\n") + "\n"
}

// Unfilled — переменные, которые человек ещё не заполнил.
func (p *ComposePlan) Unfilled() []string {
	var out []string
	for k, v := range p.Env {
		if strings.Contains(v, "<") {
			out = append(out, k)
		}
	}
	sort.Strings(out)
	return out
}

// toYAML печатает нашу же структуру.
//
// Полноценная библиотека здесь не нужна: мы выводим то, что сами построили, а
// не разбираем чужое. Строки всегда в двойных кавычках — так не нужно гадать,
// какие из них YAML примет за число, дату или значение вроде no.
func toYAML(v any, indent int) string {
	pad := strings.Repeat("  ", indent)

	switch t := v.(type) {
	case nil:
		return " null\n"

	case map[string]any:
		if len(t) == 0 {
			return " {}\n"
		}
		keys := make([]string, 0, len(t))
		for k := range t {
			keys = append(keys, k)
		}
		sort.Strings(keys)

		var b strings.Builder
		b.WriteString("\n")
		for _, k := range keys {
			b.WriteString(pad + k + ":" + toYAML(t[k], indent+1))
		}
		return b.String()

	case []any:
		if len(t) == 0 {
			return " []\n"
		}
		var b strings.Builder
		b.WriteString("\n")
		for _, item := range t {
			b.WriteString(pad + "- " + strings.TrimLeft(toYAML(item, indent+1), " "))
		}
		return b.String()

	case string:
		return " \"" + strings.NewReplacer(`\`, `\\`, `"`, `\"`).Replace(t) + "\"\n"

	default:
		return fmt.Sprintf(" %v\n", t)
	}
}
